artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/AMQPConnectionContext.java

```diff
@@ -436,16 +434,18 @@ public void onRemoteOpen(Connection connection) throws Exception {
       public void run() {
          Long rescheduleAt = handler.tick(false);
 
-         synchronized (schedulingLock) {
-            if (!isSchedulingCancelled) {
+         scheduledFutureRef.getAndUpdate(value -> {
+            if (value != null) {
                if (rescheduleAt == null) {
```

Comment:
this can be moved outside the function, as if rescheduleAt == null, then you dont need to make a getAndUpdate method call, but just a get, and then the function class can be one per parent object, instead of creating one per call.

Reply:
In the scenario in which close is executed before or concurrently onRemoteOpen, the scheduling must be prevented. Checking the current value of scheduledFutureRef is useful to avoid scheduling after close because it is used for 2 purpose: keeping the last scheduledFuture and the scheduling status.

Reply:
but rescheduled at, is calculated once before it hits into the getAndUpdate its doesnt change, either it will be null or not, before the atomic, and there for the logic causing can come outside the function and we can then make the function have not outside value purer functional.

Reply:
e.g.
```
         Long rescheduleAt = handler.tick(false);
         if (rescheduleAt == null) {
            scheduledFutureRef.getAndUpdate(value -> value == null ? null : scheduledPool.schedule(scheduleRunnable, 10, TimeUnit.MILLISECONDS));
         } else if (rescheduleAt != 0) {
            scheduledFutureRef.getAndUpdate(value -> value == null ? null : scheduledPool.schedule(scheduleRunnable, rescheduleAt - TimeUnit.NANOSECONDS.toMillis(System.nanoTime()), TimeUnit.MILLISECONDS));
         }
```